### ISSUE TITLE
ft_launcher 0.6.0 hot-spare: re-rendezvous with live agents > max_nodes permanently closes the rendezvous ('Maximum number of nodes exceeded') and kills the whole job

### ISSUE THREAD

**Issue by cxiong-ship-it** [NVIDIA/nvidia-resiliency-ext#348]
## Summary

With `ft_launcher` running a **hot-spare topology** (5 agents, `--nnodes 4`), a re-rendezvous round that starts while **all 5 agents are alive** can fatally close the rendezvous on the host node with:

```
ft_rendezvous_barrier.py:1458 [node-A...] Maximum number of nodes (4) exceeded. Active participant count would be 5 (joined=5, left=0). This is likely a configuration error - please check max_nodes setting.
```

after which **every agent exits with `RendezvousClosedError` and the whole SLURM job dies** — a fully recoverable single-node fault escalates to total job loss. The same 5-agents-alive re-rendezvous succeeded in a sister run (extra agent parked, training continued), and **initial formation always parks the extra agent correctly**, so this is a race in subsequent-round handling, not a configuration error as the message suggests.

## Environment

- nvidia-resiliency-ext **0.6.0** (`ft_launcher`, barrier rendezvous implementation — log lines from `ft_rendezvous_barrier.py`)
- PyTorch `2.12.0a0+0291f960b6.nv26.04` (NGC-based container), Python 3.12
- 5 nodes × 8 GPUs (B300, EFA fabric), SLURM + pyxis
- Megatron-LM with `--enable-ft-package` (fresh-process restart only; **no** `--inprocess-restart`)
- Launch (per node, 5 agents total):
  `ft_launcher --rdzv-backend=c10d --rdzv-endpoint=node-A:PORT --rdzv-id=$JOBID --nproc-per-node 8 --nnodes 4 --max-restarts 100 --ft-cfg-path ft.yaml ...`
- `ft.yaml`: `rank_section_timeouts: {setup: 900, step: 120, checkpointing: 420}`, `rank_out_of_section_timeout: 300`
- Intent: 4-node active world + 1 hot spare parked in the rendezvous wait/redundancy list (works as designed at job start and across earlier kill tests).

## Timeline (job 249886, 2026-06-10, all times UTC; full agent logs available)

1. **20:57:17** — round 0 forms: nodes A,B,D,E active (ranks 0–3), node C parks as spare. Training runs normally.
2. **~21:05** — fault injection: all 8 worker processes on node B (active rank 1) are SIGSTOPped (hang simulation; agent stays alive).
3. **21:08:30–34** — `step` section timeout detects the hang; agents restart the worker group. **Round 1** forms: A,B,D... actually **A (rank 0), E (rank 3), C (rank 2), B (rank 1)** — the former spare C is admitted, healthy node D is benched:
   ```
   21:08:34,193 [node-A] ft_rendezvous_barrier.py:2169 Assigned active rank 0 (round=1)
   21:08:34,215 [node-E] ft_rendezvous_barrier.py:2169 Assigned active rank 3 (round=1)
   21:08:34,237 [node-C] ft_rendezvous_barrier.py:2169 Assigned active rank 2 (round=1)
   21:08:34,273 [node-B] ft_rendezvous_barrier.py:2169 Assigned active rank 1 (round=1)
   21:08:34,731 [node-D] ft_rendezvous_barrier.py:1085 [node-D] [Step 0] Round 1 closed (training in ...
   ```
   The round-1 world successfully reloads a checkpoint and resumes training.
4. **~21:10:40** — the round-1 world hits a failure (under separate investigation) and the agents initiate another restart; workers receive SIGTERM and begin an exit-handler checkpoint save at 21:10:43.
5. **21:11:42** — **cycle-2 rendezvous starts with all 5 agents alive. The host agent (node A) raises the fatal error instead of parking the extra:**
   ```
   21:11:42,800 [node-A] profiling.py:90 - Cycle: 2 Event: rendezvous_started ...
   21:11:42,800 [node-A] ft_rendezvous_barrier.py:1458 [node-A] Maximum number of nodes (4) exceeded. Active participant count would be 5 (joined=5, left=0). This is likely a configuration error - please check max_nodes setting.
   21:11:42,898 [node-C] ft_rendezvous_barrier.py:969 The node 'node-C' detected that rendezvous was permanently closed
   21:11:43,894 [node-D] launcher.py:1618 Agent .run() raised exception, e=RendezvousClosedError(...)
   21:11:43,896 [node-D] ft_rendezvous_barrier.py:2389 The node 'node-D' has permanently closed the ...
   21:11:43,9xx [nodes B,C,D,E] launcher.py:3544 Agent run ended with exception, e=RendezvousClosedError(...)
   ```
6. All srun tasks exit; the SLURM job dies.

## Expected behavior

A re-rendezvous round with `live agents > max_nodes` should behave like initial formation: admit `max_nodes` participants and park the extras in the wait/redundancy list. That is exactly what happened in a sister run (job 249748, same topology, same hang injection): the post-restart round formed 4-of-5, the benched node parked, and training ran to completion — so 0.6.0 *can* handle it; it just doesn't reliably.

## Impact

The hot-spare pattern (allocate N+1, `--nnodes N`) is the natural way to get in-job node-swap recovery with fresh-process restart. This bug means any restart round that begins while all N+1 agents are alive (i.e., any *worker*-level fault, as opposed to a node death) can nondeterministically kill the entire job — the FT layer converts a recoverable fault into total job loss. Observed once in two such restart rounds.

## Reproduction sketch

1. Launch 5 `ft_launcher` agents with `--nnodes 4` (c10d rdzv, one agent per node, any torch-distributed workload that runs steps and heartbeats via the section API).
2. Confirm round 0 forms 4 active + 1 parked.
3. SIGSTOP all workers on one *active* node; wait for the `step` section timeout to trigger a group restart (round 1 forms from the 5 live agents; one parks).
4. Cause/wait for any subsequent restart (round 2) while all 5 agents are still alive. Some fraction of these rounds hit `ft_rendezvous_barrier.py:1458` "Maximum number of nodes exceeded" on the rdzv-host agent → `RendezvousClosedError` everywhere → job death.

## Secondary question

What triggered cycle 2 (~2.5 min after the round-1 world resumed) is still under investigation on our side — if a benched/parked agent's state can itself trigger a cluster restart in 0.6.0, that would make this failure mode much more likely and we'd appreciate any pointers.

We can share full (lightly redacted) agent logs for both the failing run and the surviving sister run on request.

**Comment by cxiong-ship-it** [comment]
Update on the "secondary question" (what triggered cycle 2 ~2.5 min after the round-1 world resumed): root-caused. The round-1 world had loaded a **replica-retrieved local checkpoint** (local checkpointing with --replication, jump=8, factor=2) and hit a deterministic `NaN in local forward loss` 3 iterations after the load (reproducible under Megatron's rerun_state_machine; identical signature at the same iteration in a second run). The rerun protocol saved a diagnostic checkpoint and exited requesting a different GPU — that worker exit is what initiated cycle 2, which then hit the fatal "Maximum number of nodes exceeded" close reported here.

Two follow-ups possibly relevant to this repo: (1) the replica retrieval/assembly path producing a state that NaNs deterministically may be a nvidia-resiliency-ext local-checkpointing bug rather than Megatron-LM (an identical run without --replication recovered cleanly through the same node-kill + spare swap); (2) the downstream no-fallback/looping behavior is filed as NVIDIA/Megatron-LM#5281. Happy to split (1) into a separate issue here if useful.

**Comment by hexinw-nvidia** [comment]
You need to specify "--nnodes 4:5" to support the hot spare node.
